### DISCUSSION TITLE
Fishery and BeeHive Function

### DISCUSSION THREAD

**Discussion by bevans2000** [mars-sim/mars-sim#319] [Ideas]
The activity of farming Fish & Bees could be separated into new dedicated `Function` classes. This would simplify the existing Farming class. It would allow the different types of Greenhouses to be more specific and varied. Or even one that was just a large Pond.
Coupled with this could be new Tasks, e.g CollectionHoney or CatchFish that could replenish the Inventory.

**Comment by Urwumpe** [comment]
Makes sense, since these are also completely different processes compared to farming plants.

**Comment by mokun** [comment]
> The activity of farming Fish & Bees could be separated into new dedicated `Function` classes. This would simplify the existing Farming class. 

Sure. @bevans2000, @Urwumpe, @scud1 ,

Do feel free to jump into this project as I didn't finish what I started.

I'd love seeing fishes and bees making their ways into our Mars colonies soon. 

>It would allow the different types of Greenhouses to be more specific and varied. Or even one that was just a large Pond.

Oh yea that's very good points that one should be able to customize greenhouses with different configurations to suit its purpose.

btw, a while back, we were thinking of algae pond but it never got implemented.

> Coupled with this could be new Tasks, e.g CollectionHoney or CatchFish that could replenish the Inventory.

Right, That would give the Cook, the Botanist, the Chemist, etc. more things to do and anyone feel like taking a break and enjoying time with lively things in a greenhouse :D

**Reply by Urwumpe** [reply]
What about also adding some non-food jobs that could be added to a green house, for example cultivating flowers or meditating/relaxing in a greenhouse?

**Reply by Urwumpe** [reply]
And... could there maybe be any building blocks or tasks to simplify creating such greenhouse actions? Like watering plants either manually or using automated systems if available?

**Reply by bevans2000** [reply]
Yes these would make sense but we need to make Tasks easier to add. Whilst doing the current work on TendFishTank (#320) it became apparent that there and numerous places a new Task has to be registered. I think Task should be registered in one place and then the framework takes over. I have added some ideas to #297 on getting the TaskMeta to be more descriptive of the Task attributes so we can make the framework data driven instead of explicit.

**Reply by mokun** [reply]
> Yes these would make sense but we need to make Tasks easier to add. 

True. Right now, we have ~60 Task class files and may be ~50 MetaTask files now. Will only continue to grow in numbers.

So we do need to discuss better ways of managing Task and MetaTask.

May be we can start with creating a template for all new and future tasks on the horizon ?

Any possibility of creating a Task Engine and use JSON to save attributes and have a few template, each for different type of tasks, etc. ?

**Comment by mokun** [comment]
> ![image](https://user-images.githubusercontent.com/1168584/110640157-5c96d180-8165-11eb-91df-8665c579a5e1.png)

@bevans2000 , @Urwumpe 

is 6347.6 kg weed too much ?  

Say if, the initial mass of the weed can be 100 kg or something, we'll need to find a scientific rate of growth of the weed.

Also, I'd say from now on, we shouldn't do everything for the player right off the bet automatically.

Right now, we give them 1002 fish (as well as each inflatable greenhouse having 5 crops) to begin with.

What if we assume each settlement has 1000 fish eggs to start with. Say, each batch has 20 eggs and there is 50 batches. 

They will need to choose when to open each batch.

We could redesign the interface to let the player to click on a button or choose from a menu before opening up a batch of fish eggs.

This way, they can feel a better sense of accomplishment.

**Comment by bevans2000** [comment]
These values were the initial scratch set I worked with. I've refined them in the last commit. The number of fish is based on a formula I found about Fish Density which is based on the volume of water. The default fish is now 132. The weed mass is based on the number of fish so this has come down to 940 kg; still too much. There is a ratio of 30 weeds per fish.

**Reply by mokun** [reply]
> The number of fish is based on a formula I found about Fish Density which is based on the volume of water. The default fish is now 132. The weed mass is based on the number of fish so this has come down to 940 kg; still too much. There is a ratio of 30 weeds per fish.

![image](https://user-images.githubusercontent.com/1168584/110922932-2bd5aa00-82d5-11eb-8d2a-c64300471396.png)

Are we allocating too much water for ONE BIG tank ?

We could start with just one small tank and let player click on a button to build extra tanks if they want to. 

Also, I'm thinking of showing the players a few *category* of fish, may be in terms of size and/or the age.

```
< 3 sols old : 5
< 7 sols old : 10
< 14 sols old : 12
< 28 sols old : 15
< 56 sols old : 5
< 28 * 4 years old : 3 
```

This way, the player will have a sense of seeing them growing.

At some point, we will need to decide how we turn the fish into a viable food source. 

May be we let player build a schedule of harvesting fish. 

e.g. how often to catch an older fish for harvest, etc.

**Reply by bevans2000** [reply]
Yes I think the tank is too big. We can easily change it size in the building.xml file. Also it won't be difficult to convert the single tank model into multiple tanks. However there needs to be some research on the numbers in terms of growth rates. The mechanism that is now present can be changed to support different rates.
In the `Fishery.catchFish` method it stores FISH_MEAT & FISH_OIL based on the fish size. You can see this from the console by using the command:
> inventory fish
Which should anything in the inventory with fish in the name. The Inventory command searches the name if an argument is used.

**Comment by mokun** [comment]
In terms of just features, we could draw some insight from this cute 2.5D game called [Megaquarium](https://store.steampowered.com/app/600480/Megaquarium/) with cartoon-looking assets. 

But we'd love to continue to have scientific modeling, in the tradition of mars-sim.

![image](https://user-images.githubusercontent.com/1168584/111011081-32076d00-834d-11eb-97b8-8f0ed2c99e31.png)

![image](https://user-images.githubusercontent.com/1168584/111011231-9296aa00-834d-11eb-9d0d-6f35ca39a92b.png)

![image](https://user-images.githubusercontent.com/1168584/111011094-3f245c00-834d-11eb-8f61-0d6b79b5a4e6.png)
